src/wg/format/png/reader.d

```diff
@@ -0,0 +1,926 @@
+                    output16[writePos++] = input16[x];
+                    output16[writePos++] = input16[x+1];
+                    output16[writePos++] = input16[x+2];
+                    if (input16[x..x + 3] == transparency16[0..3]) output16[writePos++] = 0;
+                    else output16[writePos++] = 0xffff;
+                    x += 3;
+                }
+                output16 = output16[writePos..$];
+                input16 = input16[samples..$];
+            }
+        }
+        else
+        {
+            foreach (y; 0..height)
+            {
+                int writePos = 0;
+                int x = 0;
+                while (x < samples)
+                {
+                    output[writePos++] = input[x];
+                    output[writePos++] = input[x+1];
+                    output[writePos++] = input[x+2];
+                    if (input[x..x + 2] == transparency[0..3]) output[writePos++] = 0;
+                    else output[writePos++] = 0xff;
+                    x += 3;
+                }
+                output = output[newLineBytes..$];
+                input = input[lineBytes..$];
+            }
+        }
+
+        result.rowPitch = newLineBytes;
+        return true;
+    }
+
+    private bool setupFormat() nothrow @nogc
+    {
+        static immutable string[] formats = [
+            "",
+            "l_1",
+            "l_2",
+            "l_4",
+            "l",
+            "l_16",
+            "rgb",
+            "rgb_16_16_16",
+
+            "",
+            "la_1_1",
+            "la_2_2",
+            "la_4_4",
+            "la",
+            "la_16_16",
+            "rgba",
+            "rgba_16_16_16_16",
+        ];
+
+        int intFormat;
+        import core.bitop: bsf;
+        final switch (colorType)
+        {
+            case PngColorType.grayscale:
+                intFormat = bitDepth.bsf() + 1;
+                if (transLength > 0)
+                {
+                    if (!moveGrayscaleTransparencyToAlpha()) return false;
+                    intFormat |= 8;
+                }
+                break;
+            case PngColorType.grayscaleWithAlpha:
+                intFormat = 8 | (bitDepth.bsf() + 1);
+                break;
+            case PngColorType.paletteColor:
+                intFormat = 6;
+                if (!depalettize()) return false;
+                if (transLength > 0) intFormat |= 8;
+                break;
+            case PngColorType.rgbColor:
+                intFormat = bitDepth == 8 ? 6 : 7;
+                if (transLength > 0) 
+                {
+                    if (!moveRgbTransparencyToAlpha()) return false;
+                    intFormat |= 8;
+                }
+                break;
+            case PngColorType.rgbaColor:
+                intFormat = bitDepth == 8 ? 14 : 15;
+                break;
+        }
+
+        switch (intFormat) {
+            case     1, 2, 3, 4, 5: result.bitsPerBlock = bitDepth; break;
```

Comment:
I haven't written this anywhere, tiny-pixels still aren't really implemented, but I think it will be a requirement that bitsPerBlock is a multiple of 8, and for cases like 1bpp, you should make the blocks larger until it reaches a byte-interval. otherwise any iterating code is really tricky.
so for `l_1`, do `blockWidth = 8; blockHeight = 1; bitsPerBlock = 8`, that will work well.

Reply:
...obviously you may need to pad rows of pixels in the output image data if the width is not aligned with the block size (most image formats will require this in practise anyway).

Reply:
So if I understand you correctly I leave the data as is (it is padded to a full byte anyway) and just set `blockHeight=1` and:
```
blockWidth bitsPerBlock "format"
    8           8       "l_1",
    4           8       "l_2",
    2           8       "l_4",
    1           8       "l",
    1          16       "l_16",
    1          24       "rgb",
    1          48       "rgb_16_16_16",

    4           8       "la_1_1",
    2           8       "la_2_2",
    1           8       "la_4_4",
    1          16       "la",
    1          32       "la_16_16",
    1          32       "rgba",
    1          64       "rgba_16_16_16_16",
```

Reply:
Exactly.

Reply:
I might change it to `blockBytes` rather than `bitsPerBlock`. It was an evolution of `bitsPerPixel`, but then I realised that I needed blocks at the lowest level, so it changed.